src/app/mainwindow.cpp

```diff
@@ -315,20 +315,19 @@ bool MainWindow::event(QEvent* event)
 
 void MainWindow::closeEvent(QCloseEvent* event)
 {
-    if (isVisible()) {
-        saveState();
-        d.save = false;
-
-        foreach (IrcConnection* connection, d.connections) {
-            connection->quit(qApp->property("description").toString());
-            connection->close();
-        }
+    saveState();
+    d.save = false;
 
-        // let the sessions close in the background
-        hide();
-        event->ignore();
-        QTimer::singleShot(1000, qApp, SLOT(quit()));
+    foreach (IrcConnection* connection, d.connections) {
+        connection->quit(qApp->property("description").toString());
+        connection->close();
     }
+
+    // let the sessions close in the background
+    hide();
+    QThread::sleep(1);
```

Comment:
Uh-oh, I assumed that irc connections run in separate threads. If not then sleep is wrong, I guess. And we probably shouldn't use something like waitForDisconnected here too?

Maybe just leave the timer connection as it was and also connect `IrcConnection::disconnected` here to some function that calls `qApp->quit()` once all connections are gone?

Reply:
Sure, that would also work if we set QGuiApp::quitOnLastWindowClosed to false. The reason why it was done like this with an arbitrary 1s timer was pure laziness and carelessness. "If 1s is not enough to close the connections gracefully, we don't care. In worst case it just results to 'remote closed the connection'". Surely it would be better to get the application closed right away when the connections have been closed. :)